### ISSUE TITLE
Performance problems

### ISSUE THREAD

**Issue by fremoikaner** [Extended-Thunder/send-later#248]
Hi all!
At first a big "thanks" to bring this useful extension to a new level.
I encounter massive performance issues when send-later is enabled. It uses around 12-15% CPU-usage and is slowing down more and more from hour to hour. A restart helps for a little while but after some hours it's almost impossible to work with TB.
When I disable send-later the CPU-usage drops directly to 1%. enabling it and it has again the high CPU-usage again.
As the complete mail-folder is about 90 GB(!) I started also some other helpful tipps (e.g. disabling the hardware-acceleration, etc.) and also deleted all MSF-files but the problem still persists.

My config:
OS: Win7(64-Bit), TB 78.6.0 (32-Bit), SendLater: 8.6.5
If you need some more details just let me know which.

Thanks in advance and have a Merry Christmas and a Happy New Year (hopefully better than the old one was ...)
Andreas

**Comment by jph-sendlater** [comment]
Do you mean that your Drafts folder itself is 90GB, or the entire Thunderbird profile with all associated accounts?

One performance issue I've seen while testing happens when you set the 'Console log level' to anything higher than its default ("Info"). When it's on "Debug" "Trace" or "All" there are a *lot* of messages printed to the error console, which can grow quite large. For some reason Thunderbird then starts getting progressively slower. However, that doesn't typically increase CPU load in a noticeable way-- just makes Thunderbird generally laggy in my experience.

I don't really know what would be causing this off hand, but these are some generic details that might be helpful:

* What do your Send Later preferences look like? Try opening the "Advanced configuration editor" in the Send Later options page and copy-paste its content into a comment here (you can omit the 'instanceUUID' property, since it's specific to your profile and not necessary here).
* What is your mail server / profile setup like? (IMAP or POP3?, multiple accounts?, etc)
* Any other extensions installed?
* In your account "Server Settings" page, do you have the option "When I delete a message:" set to "Just mark it as deleted"?

**Comment by fremoikaner** [comment]
Hi!
Thanks at first for the quick response.
The complete profile is around 90 GB.
The console-level is set to "Warning". I also tried other settings, but no change in CPU-usage visible.

Here's the preferences:

{
  "versionNumber": "8.2.1",
  "checkTimePref": "0",
  "checkTimePref_isMilliseconds": false,
  "quickOptions1Label": "in 15 minutes",
  "quickOptions1funcselect": "Delay",
  "quickOptions1Args": "15",
  "quickOptions2Label": "in 30 minutes",
  "quickOptions2funcselect": "Delay",
  "quickOptions2Args": "30",
  "quickOptions3Label": "in 2 hours",
  "quickOptions3funcselect": "Delay",
  "quickOptions3Args": "120",
  "showHeader": true,
  "showColumn": true,
  "showStatus": true,
  "logConsoleLevel": "warn",
  "logDumpLevel": "fatal",
  "logFilter": "",
  "sendUnsentMsgs": true,
  "sendDrafts": true,
  "sendButton": true,
  "sendDoesDelay": false,
  "sendDelay": 10,
  "sendDoesSL": true,
  "showEditAlert": false,
  "showOutboxAlert": true,
  "showSendNowAlert": false,
  "altBinding": false,
  "sendWhileOffline": true,
  "enforceTimeRestrictions": true,
  "blockLateMessages": true,
  "lateGracePeriod": 20,
  "activeScanner_lastUUID": deleted by me
  "activeScanner_UUID": "",
  "activeScanner_time": 1605622133,
  "askTime": 1473147570,
  "askSent": -2,
  "askQuit": false,
  "relNotes": "7.4.4",
  "disabledForEnigmail": false,
  "editorFirstUse": false,
  "promptDefaults": "[\"17. November 2020 17:52\",\"none between 1600 2000 on 1 2 3 4 5 6\",[],\"\"]",
  "markDraftsRead": false,
  "composeWindowHotkeys": true,
  "instanceUUID": deleted by me
  "migratedLegacy": 4,
  "throttleDelay": 0
}

I only have pop3-account (around 8) at different providers (e.g. GMX, or my own hosted domain)
Currently all other extensions are greyed out as they are not compatible with TB 78.
At the server-settings page I do not have the stated option.

What I can add is ...
with deactivated SendLater TB starts in approx. 2 seconds, prompts for the master-password which can be entered immediately and opens afterwards the mailbox after 10 seconds.
When SendLater is active and I start TB I also get the prompt after 2 seconds, but it takes around 10 more seconds until I can enter the master-password. During this period I also see in the header "no response" after that the behaviour is identical.

Thanks for your support :-)
Andreas

**Comment by jph-sendlater** [comment]
> At the server-settings page I do not have the stated option.

Do you mean you don't have the option set to "mark as deleted" or you don't have the option available to you at all?

Did you intend to set your "check every ___ minutes" option to 0? The effect of that is to disable Send Later from sending messages. In a way that makes this more mysterious, since Send Later should be basically idle all the time.

Try opening the error console (Ctrl+Shift+J) and changing the "console log level" to "All" (remember to change it back when you're done). It might become obvious right away if Send Later is stuck in some sort of infinite loop.

**Comment by fremoikaner** [comment]
I don't even have this option. I will send a screenshot of the settings and the log-file directly to you via e-mail as I don't know if it might contain confidential information.
Yes, the check-value is set to 0 as I currently don't want SendLater to send any mails as officially my business is closed until 2021.
Thanks!

**Comment by jph-sendlater** [comment]
I didn't see anything obvious in your logs, but it is notable that the 'Entering main loop' message never shows up. It seems like something must be blocking that function from executing, or some error is not being properly caught and reported.

Your log did lead me to find one small bug, which is fixed in 5629ff7f29c25e2e419279b221dcbe9b94c9c89d, but I don't expect that caused the performance problems you're having.

For right now, I think I'll recommend holding off on making any changes until your business is back open and you can more freely test different options. The next release of Send Later (v8.7.1) will be available as soon as the Thunderbird reviewers are back from their holiday, and we can do more thorough debugging if the problem persists with that version.

**Comment by fremoikaner** [comment]
Hi Jonathan,
thanks for checking the files and settings :-)
When the new version (8.7.x) will be officially available I will uninstall the old one complete, restart TB with Admin-rights and install the new version. I will report about the results here.
What I figured out during the last days is ... that the cpu-usage of ~13% (which is 1/8 of 100%) is stable when sending an e-mail with activated SendLater. This holds almost exactly 9 1/2 minutes then cpu-usage drops to almost zero.
measured times: 10:08, 10:02, 9:21, 9:47, 9:36 minutes.

After that I sent out two mails (with send later set to 04-01-2021) almost at the same time and the time of high cpu-usage almost doubles (19:23 and 20:32 minutes).
I have no idea if this helps to find a reason - just noticed it ;-)

Happy New Year!
Andreas

**Comment by jph-sendlater** [comment]
version 8.7.1 should be available now. Give that a shot when you have a chance.

**Comment by fremoikaner** [comment]
Hi again,

sorry for the delay ...
Yesterday I took an hour to work on these sendlater-issues.

It updated automatically to 8.7.2. After a restart of TB I had the same performance-issues as already reported.
So I re-started TB with Admin-rights (usually I work as user only) and uninstalled SendLater completly.
After the next restart TB had just during the start-phase a higher cpu-usage, after a few seconds it drops to almost zero.
Sending out a test-mail worked quite well.
So I decided to install SendLater again and restartet TB again. As described before already the message-box to enter the master-password took approx. 10 seconds with "no respone" - after that I could enter the password and continue.
Some seconds later TB is fully started and cpu-usage is again at ~13% - stable. TB runs quite slow when typing, moving mails or folders, etc. This holds on until I disable TB - then cpu-usage drops immediately to zero.
Currently there is no other extension installed anymore - I removed all of them (except SendLater).

I also enabled it again 10 minutes prior to the frist scheduled e-mails and waited until the time came. The interval in SendLater is set to 1 minute. But nothing happened. Not in the minute when the messages should be send, not in the minutes afterwards.
So i disabled it again and sent out the mails directly. This is reproducable with other time-slots as well - so currently SendLater is a No-send ;-)

As by business is running again regularly, we can test whatever might be helpful. Currently I have almost no scheduled mails in the draft-folder.
Might it help to delete the draft-folder and re-start TB? Will it be created automatically new?
I also created two screenshots showing the status-messages - if this helps, please let me know to whom I can send them as they contain customer-mail-addresses.

Cheers
Andreas

**Comment by fremoikaner** [comment]
Maybe as addition I write what will be visible in the sreenshots:
The first one shows the message in the status bar with "Send later [checking for messages to send]" (maybe the wording is a bit different - I see the german translation there). This status remains for a long time-period.
I remember in one of the first SendLater installations that it goes for just a second or so in this status.

The second one shows the time of the PC (17:27) and the send-later times of some mails in the draft-folder (17:10) - of course on the same day (05-01-2021).
The status bar this time is "Send later [pending: 21]" (re-translated from the german text).

**Comment by jph-sendlater** [comment]
Usually the "checking for messages" status only lasts for a few seconds at most, so that might be an important observation. That would suggest that Send Later is scanning folders that don't need to be scanned -- although it should have been skipping that altogether when you had checkEvery set to "0", so that doesn't quite add up.

Would you mind doing another debug log now that you have re-enabled sending (by setting check every __ minutes > 0)? You can send it to [send-later-support@extended-thunder.org](mailto:send-later-support@extended-thunder.org) -- the previous one went to Jonathan Kamens, who is no longer working on Send Later, although it was also forwarded to me.

**Comment by fremoikaner** [comment]
Thanks again for your support :-)
I create a new log-file and sent it to the stated e-mail-address.

Cheers
Andreas

**Comment by jph-sendlater** [comment]
@fremoikaner are you still experiencing this issue (in the latest version: 8.9.2)? I've changed a lot of things in Send Later since this bug was opened.

**Comment by jph-sendlater** [comment]
I'm going to close this for now, assuming it's no longer an issue. Let me know if that isn't the case and I can open it back up.

**Comment by fremoikaner** [comment]
Sorry, I answered via e-mail - maybe it did not reach you ...

sorry for answering late, but I wanted to give it a new chance
I closed TB, restarted it, activated SendLater again - and have the same issues like before.
It slows down more and more also the typing, the handling, etc.
Also it takes again the 13% cpu-usage like before.

I checked the folder "drafts" and it has the "send later" column, but it is empty.

I also tried to send out some mails with SendLater e.g. +15 minutes - but it did not send them at the desired time.
After +20 minutes after the estimated sending time I receive the message, that the sending is delayed ...

You wrote that you have/had another client with the same issues - were you able to solve his problem in the meantime?

Can I help in any case with files, screenshots, settings?

Best regards/Viele Gruesse
fremoikaner
